### PR TITLE
Change to operates_on metadata in LFRic

### LINKED ISSUE #866: [i-first] Change kernel meta-data from "iterates_over" to "operates_on"
For phase I of i-first support for LFRic we need support for a new type of kernel. This kernel (which maps from LFRic to UM Physics kernels) will need to be supplied with entire LFRic fields - as opposed to being called from within a loop over columns or dofs.

### PR DESCRIPTION
**Pull request by arporter**
This goes some way towards #866 - it provides all the functionality and updates some of the test kernels. However, it does *not* change the metadata in all of the test kernels and examples. I'll do that as a separate PR.

### CODE REVIEW

**Comment by arporter** [comment]
This is now ready for first review from either @rupertford or @TeranIvy

**Comment by TeranIvy** [comment]
Taken.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 1]
```diff
@@ -276,10 +276,12 @@ def test_ad_field_init_wrong_type():
 
 
 def test_ad_field_init_wrong_iteration_space():
-    ''' Test that an error is raised if a wrong iteration
-    space (other than ['cells', 'dofs']) is passed to the
+    ''' Test that an error is raised if a wrong iteration space
+    (other than ['cell_column', 'dof']) is passed to the
     LFRicArgDescriptor._init_field() method.
 
+    TODO #870 update this test.
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_test.py:283]: Please add a bit more information on how specifically this test needs to be updated.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 2]
```diff
@@ -290,9 +292,10 @@ def test_ad_field_init_wrong_iteration_space():
         LFRicArgDescriptor(
             field_arg, metadata.iterates_over)._init_field(
                 field_arg, "ncolours")
-    assert ("LFRicArgDescriptor._init_field(): Invalid iteration "
-            "space 'ncolours' in the kernel metadata (expected "
-            "one of ['cells', 'dofs'])." in str(excinfo.value))
+    assert ("LFRicArgDescriptor._init_field(): Invalid operates-on "
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_test.py:295]: As commented in `LFRicArgDescriptor`, I suggest removing the routine name and replacing `operates-on` with `operatates_on`.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 3]
```diff
@@ -396,9 +399,9 @@ def test_ad_invalid_iteration_space():
     arg_type = field_descriptor._arg_type
     with pytest.raises(InternalError) as excinfo:
         _ = LFRicArgDescriptor(arg_type, "colours")
-    assert ("LFRicArgDescriptor.__init__(): Expected one of "
-            "['cells', 'dofs'] iteration spaces in the kernel "
-            "metadata but got 'colours'." in str(excinfo.value))
+    assert ("LFRicArgDescriptor.__init__(): Expected operates_on in the "
```
Comment (TeranIvy) [inline src/psyclone/tests/dynamo0p3_test.py:402]: As above, I suggest removing the routine name for this internal error.

File: src/psyclone/tests/parse/kernel_test.py [hunk 4]
```diff
@@ -52,8 +52,11 @@
 # pylint: disable=invalid-name
 
 # Code fragment for testing standard kernel setup with
-# a type-bound procedure.
-CODE = (
+# a type-bound procedure. This uses the 'iterates_over'
+# member rather than 'operates_on'.
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:56]: What is "member"? Why not "metadata element" or just "metadata"?

File: src/psyclone/tests/parse/kernel_test.py [hunk 5]
```diff
@@ -478,13 +496,61 @@ def test_kerneltype_nargs():
...
+    assert repr(tmp) == "KernelType(test_type, cell_column)"
+
+    # With iterates_over set
+    parse_tree = parse(ITERATES_OVER_CODE)
+
     tmp = KernelType(parse_tree)
     assert repr(tmp) == "KernelType(test_type, cells)"
 
 
+@pytest.mark.parametrize('operates', ["cell_column", "dof"])
```
Comment (TeranIvy) [inline src/psyclone/tests/parse/kernel_test.py:512]: Nice use of parameterized test, I'll keep this in mind!

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Nice work! The changes are nicely contained within a limited portion of the code and I agree with the developer about establishing mapping in `dynamo0p3.py` and keeping support for both `iterates_over` and `operates_on` in `kernel.py`. 

- Tests with compilation pass.
- Changes are `pycodestyle` and `pylint` clean.
- Documentation builds.
- All changes are covered by tests.
- LFRic builds and runs (all green including the KGO checks) in the full Rose test suite with PSyclone built from this branch and metadata changed to `operates_on` in a portion of kernels in all apps. 

I have some minor remarks inline and I have a couple of requests.
1. I noticed that the LFRic built-ins metadata are not changed in this PR. I would kindly ask the developer to modify at least a couple (for instance, `setval_c` and `setval_X` as the most frequent) so we can be sure that LFRic indeed builds and runs successfully with this change - I am happy to re-test.
2. Are the metadata for all remaining LFRic examples and test kernels going to be changed in #870? There are quite a few of them so I suggest associating at least one other PR with issue #866 for the metadata change. Having to change the majority of metadata just before the release could be tricky, as there will be other significant kernel metadata changes.

File: examples/lfric/eg3/matrix_vector_mm_mod.F90 [hunk 6]
```diff
@@ -68,7 +68,7 @@ module matrix_vector_mm_mod
        arg_type(GH_FIELD,    GH_READ, ANY_SPACE_1),                    &
        arg_type(GH_OPERATOR, GH_READ, ANY_SPACE_1, ANY_SPACE_1)        &
        /)
-  integer :: iterates_over = CELLS
+  integer :: operates_on = CELL_COLUMN
```
Comment (TeranIvy) [inline examples/lfric/eg3/matrix_vector_mm_mod.F90:71]: Is the name changed from `CELL` to `CELL_COLUMN` in the `use argument_mod` statement? It doesn't seem so.

File: examples/lfric/eg3/w3_solver_kernel_mod.F90 [hunk 7]
```diff
@@ -75,7 +75,7 @@ module w3_solver_kernel_mod
        func_type(Wchi, GH_DIFF_BASIS)        &
        /)
   integer :: gh_shape = GH_QUADRATURE_XYoZ
-  integer :: iterates_over = CELLS
+  integer :: operates_on = CELL_COLUMN
```
Comment (TeranIvy) [inline examples/lfric/eg3/w3_solver_kernel_mod.F90:78]: Is the name changed from `CELL` to `CELL_COLUMN` in the `use argument_mod` statement? It doesn't seem so.

**Comment by TeranIvy** [comment]
I noticed two more things to change, @arporter, they are listed in the recent comments.

File: doc/developer_guide/APIs.rst [hunk 8]
```diff
@@ -206,10 +206,11 @@ vertical "column".
 Cells
 -----
 
-The Dynamo0.3 API currently assumes that all kernels which support
-iterating over cells work internally on a column of cells. This means
+In the LFRic API, kernels which have metadata which specifies
+``operates_on = CELL_COLUMN`` work
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developer_guide/APIs.rst:210]: Done.

File: doc/developer_guide/APIs.rst [hunk 9]
```diff
@@ -206,10 +206,11 @@ vertical "column".
...
+internally on a column of cells. This means
 that PSyclone need only be concerned with iterating over cell-columns
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developer_guide/APIs.rst:212]: Sure.

File: doc/user_guide/dynamo0p3.rst [hunk 10]
```diff
@@ -418,8 +418,10 @@ types.
 
 1) A Kernel must have at least one argument that is a field, field
    vector, or operator. This rule reflects the fact that a Kernel
-   iterates over a space and therefore must have some representation
-   over that space.
+   operates on some sub-set of the whole domain (e.g. a cell column)
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:421]: Done.

File: doc/user_guide/dynamo0p3.rst [hunk 11]
```diff
@@ -418,8 +418,10 @@ types.
...
+   and is therefore designed to be called from within a loop that
+   iterates over those sub-sets of the domain. It must therefore have
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:423]: I've had a go at re-wording but I'm not sure it's much better. Possibly that last sentence is just stating the obvious anyway and we could remove it?

File: doc/user_guide/dynamo0p3.rst [hunk 12]
```diff
@@ -1311,17 +1314,28 @@ Note that it is an error for kernel metadata to specify a value for
...
 specifies that the Kernel has been written with the assumption that it
-is iterating over the specified entity. For user-supplied kernels this
-currently only has one valid value which is ``CELLS``.
-
-Procedure
+is supplied with the specified data for each field/operator argument.
+For user-supplied kernels this currently only has one valid value
+which is ``CELL_COLUMN``. i.e. the kernel expects to be passed the
+data for a single column of cells for each field or operator argument.
+The possible values for ``OPERATES_ON`` and their interpretation is
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:1328]: Oops.

File: doc/user_guide/dynamo0p3.rst [hunk 13]
```diff
@@ -1311,17 +1314,28 @@ Note that it is an error for kernel metadata to specify a value for
...
-The fourth type of metadata provided is ``ITERATES_OVER``. This
+The fourth type of metadata provided is ``OPERATES_ON``. This
 specifies that the Kernel has been written with the assumption that it
-is iterating over the specified entity. For user-supplied kernels this
-currently only has one valid value which is ``CELLS``.
-
-Procedure
+is supplied with the specified data for each field/operator argument.
+For user-supplied kernels this currently only has one valid value
+which is ``CELL_COLUMN``. i.e. the kernel expects to be passed the
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:1326]: It was supposed to start a new sentence but I've googled and I see that I'm not supposed to do that so I've changed it :-)

File: doc/user_guide/dynamo0p3.rst [hunk 14]
```diff
@@ -1311,17 +1314,28 @@ Note that it is an error for kernel metadata to specify a value for
...
+data for a single column of cells for each field or operator argument.
+The possible values for ``OPERATES_ON`` and their interpretation is
+summarised in the following table:
+
+===========  =========================================================
+operates_on  Data passed for each field/operator argument
+===========  =========================================================
+cell_column  Single column of cells
+dof          Single dof (currently :ref:`built-ins` only)
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:1335]: Done.

File: doc/user_guide/kernel_layer.rst [hunk 15]
```diff
@@ -4,15 +4,15 @@ Kernel layer
 ============
 
 In the PSyKAl separation of concerns, Kernel code (code which is
-created to run within the Kernel layer), works over a subset of a
-field (such as a column). The reason for doing this is that it gives
+created to run within the Kernel layer), operates on a subset of a
+field (such as a column of cells). The reason for doing this is that it gives
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/kernel_layer.rst:8]: Done.

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 16]
```diff
@@ -63,18 +63,18 @@ class LFRicArgDescriptor(Descriptor):
...
 
     :raises ParseError: if a 'meta_arg' entry is not of 'arg_type' type.
     :raises ParseError: if a 'meta_arg' entry has fewer than 2 args.
     :raises ParseError: if an argument type is not one of LFRic API \
                         valid argument types.
     :raises ParseError: if the second 'meta_arg' entry is not a valid \
                         access descriptor.
-    :raises InternalError: if the iteration space from the parsed kernel \
-                           metadata is not over 'cells' or 'dofs'.
+    :raises InternalError: if the operates-on from the parsed kernel \
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py:76]: Oh yes.

File: src/psyclone/dynamo0p3.py [hunk 17]
```diff
@@ -2122,8 +2131,8 @@ def __init__(self, node):
         self._unique_indirection_maps = OrderedDict()
 
         for call in self._calls:
-            # We only need a dofmap if the kernel iterates over cells
-            if call.iterates_over == "cells":
+            # We only need a dofmap if the kernel iterates over cell_column
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py:2134]: Oh yes, done.

File: src/psyclone/dynamo0p3.py [hunk 18]
```diff
@@ -7133,15 +7141,15 @@ def gen_stub(self):
         :returns: root of fparser1 AST for the stub routine.
         :rtype: :py:class:`fparser.one.block_statements.Module`
 
-        :raises InternalError: if a kernel stub does not have a supported \
-                               iteration space (currently only "cells").
+        :raises InternalError: if the supplied kernel stub does not operate \
+                        on a supported quantity (currently only "cell_column").
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py:7145]: Oh yes, done.

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 19]
```diff
@@ -422,18 +422,18 @@ def _init_field(self, arg_type, iterates_over):
...
                            self._function_space1.lower(), arg_type))
-        # Raise an InternalError for an invalid iteration space
+        # Raise an InternalError for an invalid value of operates-on
         else:
             from psyclone.dynamo0p3 import VALID_ITERATION_SPACES
             raise InternalError(
-                "LFRicArgDescriptor._init_field(): Invalid iteration "
-                "space '{0}' in the kernel metadata (expected one of {1}).".
-                format(iterates_over, VALID_ITERATION_SPACES))
+                "LFRicArgDescriptor._init_field(): Invalid operates-on '{0}' "
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py:434]: Done and done.

File: src/psyclone/dynamo0p3.py [hunk 20]
```diff
@@ -7133,15 +7141,15 @@ def gen_stub(self):
...
 
         '''
-        # Check iteration space before generating code
+        # Check operates-on (iteration space) before generating code
         if self.iterates_over not in USER_KERNEL_ITERATION_SPACES:
             raise InternalError(
-                "DynKern.gen_stub(): Expected one of {0} as iteration "
-                "space but found '{1}' in kernel '{2}'.".
+                "DynKern.gen_stub(): Expected the kernel to operate on one of "
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/dynamo0p3.py:7151]: Done.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/tests/dynamo0p3_test.py:283]: Done (although, on reflection, this comment isn't really necessary as this test will just fail when #870 is done).

File: src/psyclone/tests/parse/kernel_test.py (hunk 4 above)
Comment (arporter) [reply src/psyclone/tests/parse/kernel_test.py:56]: Done.

File: examples/lfric/eg3/matrix_vector_mm_mod.F90 (hunk 6 above)
Comment (arporter) [reply examples/lfric/eg3/matrix_vector_mm_mod.F90:71]: Oops, done.

File: examples/lfric/eg3/w3_solver_kernel_mod.F90 (hunk 7 above)
Comment (arporter) [reply examples/lfric/eg3/w3_solver_kernel_mod.F90:78]: Done.

**Comment by arporter** [comment]
This PR only partially addresses #866. I'll do a second PR to update the metadata of the majority of LFRic kernels in the repo. Thanks for spotting that I'd not updated any built-ins - I've done a couple of them as you suggested.
Ready for second review (assuming Travis is happy).

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 21]
```diff
@@ -193,16 +193,16 @@ def __init__(self, arg_type, iterates_over):
         # Check for the allowed iteration spaces from the parsed kernel
         # metadata
         from psyclone.dynamo0p3 import VALID_ITERATION_SPACES
-        if iterates_over not in VALID_ITERATION_SPACES:
+        if operates_on not in VALID_ITERATION_SPACES:
             raise InternalError(
-                "LFRicArgDescriptor.__init__(): Expected one of {0} "
-                "iteration spaces in the kernel metadata but got "
-                "'{1}'.".format(VALID_ITERATION_SPACES, iterates_over))
+                "LFRicArgDescriptor.__init__(): Expected operates_on in the "
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py]: Thanks!

File: doc/user_guide/dynamo0p3.rst [hunk 22]
```diff
@@ -2556,7 +2571,7 @@ Configuration
 Annexed DoFs
 ++++++++++++
 
-When a kernel iterates over DoFs (rather than cells) for a continuous
+When a kernel operates on DoFs (rather than cell columns) for a continuous
```
Comment (TeranIvy) [inline doc/user_guide/dynamo0p3.rst:2574]: "cell-columns"

File: doc/user_guide/dynamo0p3.rst [hunk 23]
```diff
@@ -2566,7 +2581,7 @@ processors will have continuous fields which contain DoFs that the
 processor does not own. These unowned DoFs are called `annexed` in the
 Dynamo0.3 API and are a separate, but related, concept to field halos.
 
-When a kernel that iterates over cells needs to read a continuous
+When a kernel that operates on a cell column needs to read a continuous
```
Comment (TeranIvy) [inline doc/user_guide/dynamo0p3.rst:2584]: "cell-column"

File: doc/user_guide/dynamo0p3.rst [hunk 24]
```diff
@@ -421,8 +421,11 @@ types.
 
 1) A Kernel must have at least one argument that is a field, field
    vector, or operator. This rule reflects the fact that a Kernel
-   iterates over a space and therefore must have some representation
-   over that space.
+   operates on some subset of the whole domain (e.g. a cell-column)
+   and is therefore designed to be called from within a loop that
+   iterates over those subsets of the domain. This then only makes
```
Comment (TeranIvy) [inline doc/user_guide/dynamo0p3.rst:426]: I think you're right about this sentence just stating the obvious and removing it. Perhaps we will need further explanantion when we have more "iteration domains" with the Physics kernels.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Thanks for the changes! I have a few small comments (inline) but other than that we're ready to go.

- Tests with compilation pass.
- Changes are `pycodestyle` and `pylint` clean.
- Documentation builds (both user and developer guide) and reads well (PDF and HTML format).
- All changes are covered by tests.
- As in the previous review, LFRic builds and runs in the full Rose test suite on Spice with PSyclone built from the updated branch.

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 25]
```diff
@@ -340,17 +340,17 @@ def _init_field(self, arg_type, iterates_over):
...
-                            passed to a kernel that iterates over cells \
+                            passed to a kernel that operates on cell-columns \
                             does not have a valid access (one of \
                             [READ, WRITE, READWRITE]).
         :raises ParseError: if a field on a continuous function space \
-                            passed to a kernel that iterates over cells \
+                            passed to a kernel that operates on cell-columns \
                             does not have a valid access (one of [READ, INC]).
-        :raises InternalError: if an invalid iteration space is passed in.
+        :raises InternalError: if an invalid operates-on is passed in.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py:353]: Done (although I think its meaning is clear).

File: doc/user_guide/dynamo0p3.rst (hunk 22 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:2574]: Done.

File: doc/user_guide/dynamo0p3.rst (hunk 23 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:2584]: Done.

File: doc/user_guide/dynamo0p3.rst (hunk 24 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:426]: Removed :-)

**Comment by arporter** [comment]
I think this is ready for next review...

**Review by TeranIvy — APPROVED** [review]
All well now. Proceeding to updating the branch and merging to master.

**Pull request merged**